force-app/main/default/classes/ContactController.cls

```diff
@@ -1,170 +1,132 @@
+                        e.setMessage('first error: DELETE_FAILED, Your attempt to delete this contact could not be completed because it is associated with some cases');
+                        throw e;
+                    } else {
+                        Contact newExecutor = getNewExecutor(colleagues, stayContactsWithCases);
+                        for(Case cs : delContactsWithCases.get(c)){
+                            cs.ContactId = newExecutor.Id;
+                            updateCases.add(cs);
                         }
-                        delCases.addAll(closedCases);
-                        System.debug('delCases: ' + delCases);
-                        delContact = new Contact(Id = contactId);                 
-                        delContacts.add(delContact); 
                     }
-                }  
-                if(!delCases.isEmpty()){
-                    delete new List<Case>(delCases);
-                }
-                if (!updatedCases.isEmpty()) {
-                    update new List<Case>(updatedCases);
-                } 
-                delete delContact;                      
-            }                       
-            return delContacts;
-        } catch (Exception e) {
-            throw new AuraHandledException(' delete Contacts' + e.getMessage() + e.getStackTraceString());
-        }    
-    }
-    
-    @AuraEnabled
-    public static List<Contact> findContactColleagues(Id contactId){
-        try {
-            Account contactAccount = findAccountByContactId(contactId);
-            return [
-                SELECT Id FROM Contact WHERE AccountId = :contactAccount.Id AND Id != :contactId
-            ];          
+                }                
+            } 
+            update updateCases;
+            delete source; 
+            maxCases = 0;    
+            return source;    
         } catch (Exception e) {
-            throw new AuraHandledException('findContactColleagues' + e.getMessage());
+            throw new AuraHandledException(e.getMessage());
         }
     }
-    
-    private static Boolean isContactHasUnclosedCases(List<Case> source, Id contactId){
-        Boolean result = true;
-        List<Case> unclosedCases = CaseController.findCasesByContactId(source, contactId, unclosedStatus);
-        if (unclosedCases == null) {
-            result = false;
+
+    private static List<Id> getAccountsId(List<Contact> source){
+        Set<Id> result = new Set<Id>();
+        for (Contact c : source) {
+            result.add(c.AccountId);
         }
+        return new List<Id>(result);
+    }
+
+    private static Map<Id, List<Contact>> getColleagues(List<Contact> source, List<Id> accountsId){
+        List<Contact> temp = [SELECT Id, AccountId FROM Contact c WHERE c.AccountId IN :accountsId AND Id NOT IN :source];
+        Map<Id, List<Contact>> result = new Map<Id, List<Contact>>();
+        for (Contact c : temp) {
```

Comment:
you can directly put the SOQL from 64th line instead of "temp" here, no need to keep one more list in heap